### PR TITLE
BREAKING(k8s): change CLI API

### PR DESCRIPTION
**Pull request by josedonizetti**
## Description

before: 

```
$ trivy k8s --report=summary # scan cluster
$ trivy k8s -n default --report=summary # scan namespace
$ trivy k8s  deploy/orion # scan resource
```

after:

```
$ trivy k8s --report=summary cluster # scan cluster
$ trivy k8s --report=summary all # scan namespace
$ trivy k8s  deploy/orion # scan resource
```

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/2135

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: docs/docs/kubernetes/cli/scanning.md [hunk 1]
```diff
@@ -11,7 +11,7 @@ Trivy uses your local kubectl configuration to access the API server to list art
 Scan a full cluster and generate a simple summary report:
 
 ```
-$ trivy k8s --report=summary
+$ trivy k8s --report=summary cluster
```
Comment (chen-keinan) [inline docs/docs/kubernetes/cli/scanning.md:14]:
```suggestion
$ trivy k8s cluster --report=summary
```

File: docs/docs/kubernetes/cli/scanning.md [hunk 2]
```diff
@@ -21,13 +21,13 @@ The summary report is the default. To get all of the detail the output contains,
 Filter by severity:
 
 ```
-$ trivy k8s --severity=CRITICAL --report=all
+$ trivy k8s --severity=CRITICAL --report=all cluster
```
Comment (chen-keinan) [inline docs/docs/kubernetes/cli/scanning.md:24]:
```suggestion
$ trivy k8s cluster --severity=CRITICAL --report=all
```

File: docs/docs/kubernetes/cli/scanning.md [hunk 3]
```diff
@@ -21,13 +21,13 @@ The summary report is the default. To get all of the detail the output contains,
...
 ```
 
 Scan a specific namespace:
 
 ```
-$ trivy k8s -n kube-system --report=summary
+$ trivy k8s -n kube-system --report=summary all
```
Comment (chen-keinan) [inline docs/docs/kubernetes/cli/scanning.md:30]:
```suggestion
$ trivy k8s all -n kube-system --report=summary
```

File: docs/docs/kubernetes/cli/scanning.md [hunk 4]
```diff
@@ -40,7 +40,7 @@ The supported formats are `table`, which is the default, and `json`.
 To get a JSON output on a full cluster scan:
 
 ```
-$ trivy k8s --format json -o results.json
+$ trivy k8s --format json -o results.json cluster
```
Comment (chen-keinan) [inline docs/docs/kubernetes/cli/scanning.md:43]:
```suggestion
$ trivy k8s cluster --format json -o results.json
```

File: pkg/commands/app.go [hunk 5]
```diff
@@ -814,10 +814,10 @@ func NewK8sCommand() *cli.Command {
 		Usage:   "scan kubernetes vulnerabilities and misconfigurations",
 		CustomHelpTemplate: cli.CommandHelpTemplate + `EXAMPLES:
   - cluster scanning:
-      $ trivy k8s --report summary
+      $ trivy k8s --report summary cluster
```
Comment (chen-keinan) [inline pkg/commands/app.go:817]:
```suggestion
      $ trivy k8s cluster --report summary
```

File: pkg/commands/app.go [hunk 6]
```diff
@@ -814,10 +814,10 @@ func NewK8sCommand() *cli.Command {
...
 
   - namespace scanning:
-      $ trivy k8s -n kube-system --report summary
+      $ trivy k8s -n kube-system --report summary all
```
Comment (chen-keinan) [inline pkg/commands/app.go:820]:
```suggestion
      $ trivy k8s all -n kube-system --report summary
```

**Comment by chen-keinan** [comment]
@josedonizetti I think it would be better to implement `cluster` and `all` as `trivy k8s` <b>subcommands</b> rather than args, it will make it more modular and better to maintain

**Comment by josedonizetti** [comment]
@chen-keinan @knqyf263 Let me know what you think about the changes. I'd like to know if this look good for you, before adding integration tests, to validate the CLI works.

Things to keep in mind:
- When using Subcommands with `urfave/cli` it won't print the `CustomHelpTemplate`, I looked into fixing it, but there is a PR opened on the repository already, which took over 5 months to get a review. So, to be able to show examples of how to use, I'm using `UsageText`
- I'm mixing a command structure, where the main command handle resources `trivy k8s deployment app` and we have individual subcommands for cluster and namespace `trivy k8s cluster`, `trivy k8s all`. This allows for better maintainability but also looks good, when using subcommands we can do `trivy k8s cluster --report`, without it the flags must come before the action `trivy k8s --report cluster`. And to support the API, I didn't want to have a subcommand `resource`.
- I have extracted a `option.go` for kubernetes commands, because I don't want to add an if for every subcommand we add under `k8s` on https://github.com/aquasecurity/trivy/blob/main/pkg/commands/option/artifact.go#L41 to handle parameters diffrent. 
- The main command acts like other trivy commands `trivy k8s` shows help
- `urfave/cli` subcommands don't inherit flags, so we have to redeclare it per subcomand

**Comment by josedonizetti** [comment]
_there is an effort to introduce cobra in Trivy @knqyf263 any idea when its planned, if at all?_
@chen-keinan I don't think it is planned, I saw a PR for it, but there is no work on it.

_I'm still not sure if we should have cluster and all as subcommands. Any big benefit? For now, it makes maintenance hard._

@knqyf263 do you mean subcommands vs one command parsing argument? or do you mean separating cluster/all in different actions?

**Comment by knqyf263** [comment]
>do you mean subcommands vs one command parsing argument?

I mean the first one. Do we want to have `cluster` and `all` as subcommands?

**Comment by josedonizetti** [comment]
@knqyf263 I liked the subcommands better, mostly because of the flags. Without it, users always need to pass the flag before the command, for example `trivy k8s --report summary  all` or  `trivy k8s --format json --security-checks=config cluster`, if we do `trivy k8s all --report summary` it won't work, because `urfave/cli` is based on golang builtin flags, which stop processing whenever an argument without `-` is found. 

But with subcommands this works different, because `all` and `cluster` are no longer arguments but the commands themselves. So it works when we pass flags after it, eg: `trivy k8s cluster --format json --security-checks=config`

wdyt?

**Comment by knqyf263** [comment]
But then, `trivy k8s --format json --security-checks=config cluster` no longer works unless we define flags for `k8s` and `cluster` subcommands both, right?
Also, all other commands don't work like `trivy image alpine:3.15 --format json --security-checks=config`. I think `cluster` is a target name like `alpine:3.15`. To keep consistency with other commands, it should be like `trivy k8s --format json --security-checks=config cluster`, no?

I agree flags should work even after the CLI argument. We can do it after we move to cobra, but cannot at the moment.

**Comment by josedonizetti** [comment]
@knqyf263 @chen-keinan can you review it please? Tried to address all concerns (new CLI API, improving code maintenance and still keeping the trivy way of working)

File: pkg/commands/option/artifact.go [hunk 7]
```diff
@@ -37,16 +37,11 @@ func NewArtifactOption(c *cli.Context) ArtifactOption {
...
-	if ctx.Command.Name == "kubernetes" {
-		return nil
-	}
-
 	if c.Input == "" && ctx.Args().Len() == 0 {
 		logger.Debug(`trivy requires at least 1 argument or --input option`)
 		_ = cli.ShowSubcommandHelp(ctx) // nolint: errcheck
 		os.Exit(0)
-	} else if ctx.Args().Len() > 1 {
+	} else if ctx.Args().Len() > 1 && ctx.Command.Name != "kubernetes" {
```
Comment (knqyf263) [inline pkg/commands/option/artifact.go:44]: Does k8s subcommand accept multiple commands?

File: pkg/k8s/commands/run.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+// Run runs a k8s scan
+func Run(cliCtx *cli.Context) error {
+	opt, err := cmd.InitOption(cliCtx)
+	if err != nil {
+		return xerrors.Errorf("option error: %w", err)
+	}
+
+	cluster, err := k8s.GetCluster(opt.KubernetesOption.ClusterContext)
+	if err != nil {
+		return xerrors.Errorf("get k8s cluster: %w", err)
```
Comment (chen-keinan) [inline pkg/k8s/commands/run.go:33]:
```suggestion
		return xerrors.Errorf("failed getting k8s cluster: %w", err)
```

**Comment by chen-keinan** [comment]
could add a comment on this as it is an older change:
the error below are ignored
`_, _ = fmt.Fprintln(s.Output)`
`_, _ = fmt.Fprintf(s.Output, "Summary Report for %s\n", consolidated.ClusterName)`

File: pkg/commands/option/artifact.go (hunk 7 above)
Comment (josedonizetti) [reply pkg/commands/option/artifact.go:44]: The idea is supporting resources scanning like kubectl:

eg: 

```
trivy k8s pod pi
trivy k8s deploy orion
...
```

File: docs/docs/kubernetes/cli/scanning.md [hunk 9]
```diff
@@ -25,27 +25,27 @@ The summary report is the default. To get all of the detail the output contains,
...
 # or
-$ trivy k8s --security-checks=config --report=summary
+$ trivy k8s --security-checks=config --report=summary cluster
 ```
 
 Scan a specific namespace:
 
 ```
-$ trivy k8s -n kube-system --report=summary
+$ trivy k8s -n kube-system --report=summary all
```
[the remark this answers is not in the document]
Comment (josedonizetti) [reply docs/docs/kubernetes/cli/scanning.md:42]: Yes, `all` and `cluster` should be at the end, as they are `targets`. Like we do for other trivy commands.
`trivy image --security-checks=config alpine:latest`

**Review by chen-keinan — APPROVED** [review]

File: docs/docs/kubernetes/cli/scanning.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (knqyf263) [reply docs/docs/kubernetes/cli/scanning.md:42]: We can change it after we move to cobra.

File: pkg/commands/option/artifact.go (hunk 7 above)
Comment (knqyf263) [reply pkg/commands/option/artifact.go:44]: Thanks. It makes sense.

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
I think it is a breaking change. We should mention it in the release note. @josedonizetti Could you write it? You can refer to the old release notes.

**Pull request merged**